source/extensions/filters/http/cache/http_cache.h

```diff
@@ -0,0 +1,318 @@
+  // - Otherwise, construct a RawByteRange requesting the [first,last] body bytes.
+  // Prereq: first == UINT64_MAX || first <= last
+  // Invariant: isSuffix() || firstBytePos() <= lastBytePos
+  // Examples: RawByteRange(0,4) requests the first 5 bytes.
+  //           RawByteRange(UINT64_MAX,4) requests the last 4 bytes.
+  RawByteRange(uint64_t first, uint64_t last) : first_byte_pos_(first), last_byte_pos_(last) {
+    RELEASE_ASSERT(isSuffix() || first <= last, "Illegal byte range.");
+  }
+  bool isSuffix() const { return first_byte_pos_ == UINT64_MAX; }
+  uint64_t firstBytePos() const {
+    ASSERT(!isSuffix());
+    return first_byte_pos_;
+  }
+  uint64_t lastBytePos() const {
+    ASSERT(!isSuffix());
+    return last_byte_pos_;
+  }
+  uint64_t suffixLength() const {
+    ASSERT(isSuffix());
+    return last_byte_pos_;
+  }
+
+private:
+  uint64_t first_byte_pos_;
+  uint64_t last_byte_pos_;
+};
+
+// Byte range from an HTTP request, adjusted for a known response body size, and converted from an
+// HTTP-style closed interval to a C++ style half-open interval.
+class AdjustedByteRange {
+public:
+  // Construct an AdjustedByteRange representing the [first,last) bytes in the
+  // response body. Prereq: first <= last Invariant: begin() <= end()
+  // Example: AdjustedByteRange(0,4) represents the first 4 bytes.
+  AdjustedByteRange(uint64_t first, uint64_t last) : first_(first), last_(last) {
+    ASSERT(first < last, "Illegal byte range.");
+  }
+  uint64_t begin() const { return first_; }
+  // Unlike RawByteRange, end() is one past the index of the last offset.
+  uint64_t end() const { return last_; }
+  uint64_t length() const { return last_ - first_; }
+  void trimFront(uint64_t n) {
+    RELEASE_ASSERT(n <= length(), "Attempt to trim too much from range.");
+    first_ += n;
+  }
+
+private:
+  uint64_t first_;
+  uint64_t last_;
+};
+
+inline bool operator==(const AdjustedByteRange& lhs, const AdjustedByteRange& rhs) {
+  return lhs.begin() == rhs.begin() && lhs.end() == rhs.end();
+}
+
+std::ostream& operator<<(std::ostream& os, const AdjustedByteRange& range);
+
+// Adjusts request_range_spec to fit a cached response of size content_length, putting the results
+// in response_ranges. Returns true if response_ranges is satisfiable (empty is considered
+// satisfiable, as it denotes the entire body).
+// TODO(toddmgreer) Merge/reorder ranges where appropriate.
+bool adjustByteRangeSet(std::vector<AdjustedByteRange>& response_ranges,
+                        const std::vector<RawByteRange>& request_range_spec,
+                        uint64_t content_length);
+
+// Result of a lookup operation, including cached headers and information needed
+// to serve a response based on it, or to attempt to validate.
+struct LookupResult {
+  // If cache_entry_status == Unusable, none of the other members are
+  // meaningful.
+  CacheEntryStatus cache_entry_status = CacheEntryStatus::Unusable;
```

Comment:
member variables in structs end with `_` in Envoy style guide:

https://github.com/envoyproxy/envoy/blob/master/STYLE.md#deviations-from-google-c-style-guidelines

Reply:
done.